homeassistant/components/climate/__init__.py

```diff
@@ -288,6 +300,72 @@ class ClimateEntity(Entity, cached_properties=CACHED_PROPERTIES_WITH_ATTR_):
     _attr_target_temperature: float | None = None
     _attr_temperature_unit: str
 
+    __mod_supported_features: ClimateEntityFeature = ClimateEntityFeature(0)
+
+    def __getattribute__(self, __name: str) -> Any:
+        """Get attribute.
+
+        Modify return of `supported_features` to
+        include `_mod_supported_features` if attribute is set.
+        """
+        if __name != "supported_features":
+            return super().__getattribute__(__name)
+
+        # Implements check for deprecated use of ClimateEntityFeature as int
+        _supported_features = super().__getattribute__(__name)
+        if type(_supported_features) is int:  # noqa: E721
+            new_features = ClimateEntityFeature(_supported_features)
+            self._report_deprecated_supported_features_values(new_features)
+        return _supported_features | super().__getattribute__(
+            "_ClimateEntity__mod_supported_features"
+        )
+
+    @callback
+    def add_to_platform_start(
+        self,
+        hass: HomeAssistant,
+        platform: EntityPlatform,
+        parallel_updates: asyncio.Semaphore | None,
+    ) -> None:
+        """Start adding an entity to a platform."""
+        super().add_to_platform_start(hass, platform, parallel_updates)
+
+        def _report_turn_on_off(feature: str) -> None:
+            """Log warning not implemented turn on/off feature."""
+            report_issue = self._suggest_report_issue()
+            _LOGGER.warning(
+                (
+                    "Entity %s (%s) has not implemented ClimateEntityFeature.%s supported feature"
+                    " but is implementing the %s method. Please %s"
+                ),
+                self.entity_id,
+                type(self),
+                feature,
+                feature.lower(),
+                report_issue,
+            )
+
+        # Adds ClimateEntityFeature.TURN_OFF/TURN_ON depending on service calls implemented
+        # This should be removed in 2025.1.
+        if (
+            HVACMode.OFF in self.hvac_modes
+            or type(self).async_turn_off is not ClimateEntity.async_turn_off
+            or type(self).turn_off is not ClimateEntity.turn_off
+        ):
```

Comment:
The condition means we'll log a warning for all entities which support `HVACMode.OFF`, that can't be correct?

Reply:
In the decision it says:
```
if either of the following is true:

The derived class overrides the default implementation of climate.turn_on or climate.turn_off
The required HVAC modes are supported
```
So if mode `off` is supported then we should set the off feature flag and log a warning.
The integration does not need to implement a `turn_off` method as it can be handled by `ClimateEntity.async_turn_off`.

I think there is no other way as effectively if an integration supports mode `off` it also supports turning the entity off.

However I think perhaps we should make a slightly different message in the log then.

Reply:
What you quote talks about automatically setting the new feature flag. My concern is that the implementation in this PR logs a warning regardless of if the implementation itself sets the new flags or not.

Reply:
I noticed this so that's fixed and also a test added to ensure no adding or warning if it's already there.